{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "88a6bcd0-0877-46a5-978e-1fa6de6c71ca",
   "metadata": {},
   "source": [
    "# Result V2 "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0b8f06c-5c31-4b0f-89b7-0e7d64ae66f7",
   "metadata": {},
   "source": [
    "### import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "47d90938-3284-463a-85ba-ebd936155290",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np \n",
    "\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "from plotly.subplots import make_subplots\n",
    "\n",
    "import random\n",
    "from scipy import stats\n",
    "import math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "99b78c96-7919-4181-a7c0-e21b88febef2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reading s7\n",
      "done reading\n",
      "reading s8\n",
      "done reading\n"
     ]
    }
   ],
   "source": [
    "# file from packer and biomart\n",
    "# findbiomart version\n",
    "# explain the col drop and rename\n",
    "#PRJNA13758 WS290 \n",
    "\n",
    "#file_mart=\"/home/mateo/IGE/code_nested_gene_M/pair_list/mart_export.txt\"\n",
    "#raw=pd.read_csv(file_mart , delimiter=(\",\"))\n",
    "#raw=raw.drop(columns=raw.columns[10:14])\n",
    "#raw=raw.drop(columns='Genome name') #change mart.export then delete this line \n",
    "#raw=raw.drop_duplicates()\n",
    "## keeping only the protein_coding gene\n",
    "#raw=raw[raw['Gene biotype']=='protein_coding'].reset_index(drop=True)\n",
    "#raw=raw.reset_index(drop=True)\n",
    "\n",
    "file_mart=\"/home/mateo/IGE/code_nested_gene_M/pair_list/mart_export_WS290.txt\"\n",
    "raw=pd.read_csv(file_mart , delimiter=(\",\"))\n",
    "raw=raw[raw['Gene biotype']=='protein_coding'].reset_index(drop=True)\n",
    "raw=raw.drop_duplicates()\n",
    "raw=raw.reset_index(drop=True)\n",
    "\n",
    "file_NIHMS=\"/home/mateo/IGE/code_nested_gene_M/celltype_expression/NIHMS1604359-supplement-Table_S7.pkl\"\n",
    "print('reading s7')\n",
    "exp7=pd.read_pickle(file_NIHMS)\n",
    "#which tpm do we keep ? raw, adjusted, bootstrap ?\n",
    "exp7=exp7.drop(columns=(exp7.columns[5:10]))\n",
    "exp7=exp7.drop(columns=(exp7.columns[3]))\n",
    "exp7.columns=['gene','gene_id','lineage','tpm']\n",
    "print('done reading')\n",
    "lineage=np.unique(exp7['lineage'])\n",
    "\n",
    "file_NIHMS=\"/home/mateo/IGE/code_nested_gene_M/celltype_expression/NIHMS1604359-supplement-Table_S8.pkl\"\n",
    "print('reading s8')\n",
    "exp8=pd.read_pickle(file_NIHMS)\n",
    "#which tpm do we keep ? raw, adjusted, bootstrap ?\n",
    "exp8=exp8.drop(columns=(exp8.columns[5:10]))\n",
    "exp8=exp8.drop(columns=(exp8.columns[3]))\n",
    "exp8.columns=['gene','gene_id','lineage','tpm']\n",
    "print('done reading')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "085304ad-00f9-4753-ab5b-704370b27ed5",
   "metadata": {},
   "source": [
    "## Generate data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2afc333-f9ec-4763-bb1b-e75d65dcc012",
   "metadata": {},
   "source": [
    "#### random pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "854ced1b-9618-4181-ae37-6b78cfd34738",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rand_pair_diff(raw,nb=567):\n",
    "    pair_diff=pd.DataFrame()\n",
    "    g1=[]\n",
    "    g2=[]\n",
    "    while len(g1)<nb:\n",
    "        r1=random.randint(0,len(raw)-1)\n",
    "        r2=random.randint(0,len(raw)-1)\n",
    "        if raw.at[r1,\"Chromosome/scaffold name\"]!=raw.at[r2,\"Chromosome/scaffold name\"]:\n",
    "            g1.append(raw.at[r1,\"Gene stable ID\"])\n",
    "            g2.append(raw.at[r2,\"Gene stable ID\"])\n",
    "        \n",
    "    pair_diff[\"GeneID1\"]=g1\n",
    "    pair_diff[\"GeneID2\"]=g2\n",
    "    return pair_diff    \n",
    "# df=rand_pair_diff(raw,nb=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9722eda-8b09-4671-b1bd-f4a64ef7607c",
   "metadata": {},
   "source": [
    "#### random close pairs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e8afba71-27a1-4344-bddb-4fcf9837d554",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID1</th>\n",
       "      <th>GeneID2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WBGene00011706</td>\n",
       "      <td>WBGene00206515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WBGene00007811</td>\n",
       "      <td>WBGene00007812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WBGene00044163</td>\n",
       "      <td>WBGene00184956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WBGene00007872</td>\n",
       "      <td>WBGene00007873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WBGene00007392</td>\n",
       "      <td>WBGene00219445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>562</th>\n",
       "      <td>WBGene00194928</td>\n",
       "      <td>WBGene00006935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>563</th>\n",
       "      <td>WBGene00012541</td>\n",
       "      <td>WBGene00012542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564</th>\n",
       "      <td>WBGene00008855</td>\n",
       "      <td>WBGene00008854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565</th>\n",
       "      <td>WBGene00010290</td>\n",
       "      <td>WBGene00010288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566</th>\n",
       "      <td>WBGene00005592</td>\n",
       "      <td>WBGene00007756</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>567 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GeneID1         GeneID2\n",
       "0    WBGene00011706  WBGene00206515\n",
       "1    WBGene00007811  WBGene00007812\n",
       "2    WBGene00044163  WBGene00184956\n",
       "3    WBGene00007872  WBGene00007873\n",
       "4    WBGene00007392  WBGene00219445\n",
       "..              ...             ...\n",
       "562  WBGene00194928  WBGene00006935\n",
       "563  WBGene00012541  WBGene00012542\n",
       "564  WBGene00008855  WBGene00008854\n",
       "565  WBGene00010290  WBGene00010288\n",
       "566  WBGene00005592  WBGene00007756\n",
       "\n",
       "[567 rows x 2 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Select random pair of gene from the same chromosome and close\n",
    "\n",
    "def rand_pair_close(raw,  nb=567): #orientation = conv or div\n",
    "    pair_close=pd.DataFrame()\n",
    "    g1=[]\n",
    "    g2=[]\n",
    "    \n",
    "    #divide by chr for faster calc\n",
    "    chI=raw[raw[\"Chromosome/scaffold name\"]==\"I\"]\n",
    "    chII=raw[raw[\"Chromosome/scaffold name\"]==\"II\"]\n",
    "    chIII=raw[raw[\"Chromosome/scaffold name\"]==\"III\"]\n",
    "    chIV=raw[raw[\"Chromosome/scaffold name\"]==\"IV\"]\n",
    "    chV=raw[raw[\"Chromosome/scaffold name\"]==\"V\"]\n",
    "    chX=raw[raw[\"Chromosome/scaffold name\"]==\"X\"]\n",
    "    chMtDNA=raw[raw[\"Chromosome/scaffold name\"]==\"MtDNA\"]\n",
    "    \n",
    "    while len(g1)<nb:\n",
    "        r=random.randint(0,len(raw)-1)\n",
    "        cl=look_for_close(locals()['ch'+raw.iloc[r,3]], r)\n",
    "        if len(cl)>0:\n",
    "            g1.append(raw.at[r,\"Gene stable ID\"])\n",
    "            g2.append(cl[random.randint(0, len(cl)-1)]) \n",
    "\n",
    "    pair_close[\"GeneID1\"]=g1\n",
    "    pair_close[\"GeneID2\"]=g2\n",
    "    return pair_close\n",
    "\n",
    "\n",
    "def look_for_close(ch, r, dist=1000):\n",
    "    cl=[]\n",
    "    st, en ,sens = ch.loc[r,'Gene start (bp)'], ch.loc[r,'Gene end (bp)'], ch.loc[r,'Strand']\n",
    "\n",
    "    cl1 = ch[(ch['Gene start (bp)'] > en) & (ch['Gene start (bp)'] < en+dist) & (ch['Strand'] != sens)]['Gene stable ID']\n",
    "    cl2 = ch[(ch['Gene end (bp)'] > st-dist) & (ch['Gene end (bp)'] < st) & (ch['Strand'] != sens)]['Gene stable ID']\n",
    "    cl=pd.concat([cl1,cl2]).reset_index(drop=True)\n",
    "    return cl\n",
    "    \n",
    "df=rand_pair_close(raw,567)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2da054ab-bdca-4539-816b-966e2da93166",
   "metadata": {},
   "source": [
    "#### nested pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "cd349661-02df-474e-ab84-1a2fa1ca96c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "            GeneID1         GeneID2\n",
      "0    WBGene00011706  WBGene00206515\n",
      "1    WBGene00007811  WBGene00007812\n",
      "2    WBGene00044163  WBGene00184956\n",
      "3    WBGene00007872  WBGene00007873\n",
      "4    WBGene00007392  WBGene00219445\n",
      "..              ...             ...\n",
      "562  WBGene00194928  WBGene00006935\n",
      "563  WBGene00012541  WBGene00012542\n",
      "564  WBGene00008855  WBGene00008854\n",
      "565  WBGene00010290  WBGene00010288\n",
      "566  WBGene00005592  WBGene00007756\n",
      "\n",
      "[567 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "### select nested pairs\n",
    "\n",
    "def nested_pair(raw, name=False):\n",
    "    pair_nested=pd.DataFrame()\n",
    "    g1=[]\n",
    "    g2=[]\n",
    "\n",
    "    # separation by chr\n",
    "    chI=raw[raw[\"Chromosome/scaffold name\"]==\"I\"]\n",
    "    chII=raw[raw[\"Chromosome/scaffold name\"]==\"II\"]\n",
    "    chIII=raw[raw[\"Chromosome/scaffold name\"]==\"III\"]\n",
    "    chIV=raw[raw[\"Chromosome/scaffold name\"]==\"IV\"]\n",
    "    chV=raw[raw[\"Chromosome/scaffold name\"]==\"V\"]\n",
    "    chX=raw[raw[\"Chromosome/scaffold name\"]==\"X\"]\n",
    "    chMtDNA=raw[raw[\"Chromosome/scaffold name\"]==\"MtDNA\"]\n",
    "    #list for iteration\n",
    "    ls_chr=[chI,chII,chIII,chIV,chV,chX,chMtDNA]\n",
    "    \n",
    "    for ch in ls_chr:    \n",
    "        # since we only interessed in 1/-1 nested paire\n",
    "        strandpos=ch[ch[\"Strand\"]==1]\n",
    "        strandneg=ch[ch[\"Strand\"]==-1]  \n",
    "        \n",
    "        for i,k in strandpos.iloc[:,4:6].iterrows():\n",
    "            #start and end of the potential host gene\n",
    "            st,en=k[\"Gene start (bp)\"],k[\"Gene end (bp)\"]\n",
    "            \n",
    "            #listing every gene that start after and end before the host\n",
    "            l1=strandneg.iloc[:,4] > st\n",
    "            l2=strandneg.iloc[:,5] < en\n",
    "            \n",
    "            group=pd.concat([l1,l2],axis=1)\n",
    "            group['both']=group.sum(axis=1)\n",
    "            \n",
    "            ls_id=group[group['both']==2].index.tolist()\n",
    "            \n",
    "            #for different verbose\n",
    "            if name:\n",
    "                # for mutilpe nested in the same host\n",
    "                for j in ls_id:\n",
    "                    g1.append(ch.loc[i,'Gene name'])\n",
    "                    g2.append(ch.loc[j,'Gene name'])\n",
    "            else:\n",
    "                for j in ls_id:\n",
    "                    g1.append(ch.loc[i,'Gene stable ID'])\n",
    "                    g2.append(ch.loc[j,'Gene stable ID'])\n",
    "                \n",
    "        # same but looking for host in -1 strand            \n",
    "        for i,k in strandneg.iloc[:,4:6].iterrows():\n",
    "            #start and end of the potential host gene\n",
    "            st,en=k[\"Gene start (bp)\"],k[\"Gene end (bp)\"]\n",
    "        \n",
    "            #listing every gene that start after and end before the host\n",
    "            l1=strandpos.iloc[:,4] > st\n",
    "            l2=strandpos.iloc[:,5] < en\n",
    "            \n",
    "            group=pd.concat([l1,l2],axis=1)\n",
    "            group['both']=group.sum(axis=1)\n",
    "            \n",
    "            ls_id=group[group['both']==2].index.tolist()\n",
    "            \n",
    "            if name:\n",
    "                # for mutilpe nested in the same host\n",
    "                for j in ls_id:\n",
    "                    g1.append(ch.loc[i,'Gene name'])\n",
    "                    g2.append(ch.loc[j,'Gene name'])\n",
    "            else:\n",
    "                for j in ls_id:\n",
    "                    g1.append(ch.loc[i,'Gene stable ID'])\n",
    "                    g2.append(ch.loc[j,'Gene stable ID'])\n",
    "        \n",
    "    pair_nested[\"Host\"]=g1\n",
    "    pair_nested[\"Nested\"]=g2\n",
    "    return pair_nested.drop_duplicates(ignore_index=True)\n",
    "    \n",
    "#df=nested_pair(raw,name=(True))\n",
    "print(df)\n",
    "#df.to_csv('nested_pair_name.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c6f85938-8062-42dd-b97d-70edf373ed2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df=rand_pair_diff(raw,nb=566)\n",
    "# df=rand_pair_close(raw,'conv',567)\n",
    "# df=nested_pair(raw,name=(True))\n",
    "\n",
    "# df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d38049a3-f064-4143-8c05-0c16ca885e8a",
   "metadata": {},
   "source": [
    "#### get expression value linked to pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "5fe73545-82af-4c2f-bc3b-89e3b3d4d9d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def nb_cell_expression(df, pl, name=False):\n",
    "\n",
    "    #init the output dataframe\n",
    "    nb_cell_exp=pd.DataFrame(columns=['Host_gene','Nested_gene','exp_both','exp_U','ratio_coexp','exp_host','exp_nested'])\n",
    "\n",
    "    #Choose whether to use gene name or Whormbase ID \n",
    "    if name:\n",
    "        marker=df['gene'].drop_duplicates()\n",
    "        n_cell=len(df['lineage'].drop_duplicates()) #how many cell are in the data\n",
    "    else:\n",
    "        marker=df['gene_id'].drop_duplicates()\n",
    "        n_cell=len(df['lineage'].drop_duplicates())\n",
    "\n",
    "    # for each gene in each pair get the number of cell where tpm is above 0 \n",
    "    for k in pl.itertuples():\n",
    "        if len(marker[marker==k[1]]) > 0:            \n",
    "            i1=marker[marker==k[1]].index[0]\n",
    "            g1=df.iloc[ i1: i1+n_cell, : ].reset_index(drop=True)['tpm'] > 0\n",
    "        else :\n",
    "            g1=pd.Series([],dtype=bool) #to avoid conflict when the gene is not presente in expression data\n",
    "        \n",
    "        if len(marker[marker==k[2]]) > 0:            \n",
    "            i2=marker[marker==k[2]].index[0]\n",
    "            g2=df.iloc[ i2 : i2+n_cell, : ].reset_index(drop=True)['tpm'] > 0\n",
    "        else :\n",
    "            g2=pd.Series([],dtype=bool)\n",
    "\n",
    "        # look in how many cell both gene are expressed\n",
    "        comb=pd.concat([g1,g2],axis=1,keys=[k[1],k[2]])\n",
    "        comb['both']=comb.sum(axis=1)\n",
    "        \n",
    "        nb_cell_exp.at[k[0],'Host_gene'] = k[1]\n",
    "        nb_cell_exp.at[k[0],'Nested_gene'] = k[2]\n",
    "        nb_cell_exp.at[k[0],'exp_both'] = (comb['both']==2).sum() # get gene1 inter gene2\n",
    "        nb_cell_exp.at[k[0],'exp_U'] = g1.sum() + g2.sum() - nb_cell_exp.at[k[0],'exp_both'] # get gene1 union gene2\n",
    "        if nb_cell_exp.at[k[0],'exp_U'] == 0:\n",
    "            nb_cell_exp.at[k[0],'ratio_coexp'] = 0 # to avoid divid by 0\n",
    "        else:\n",
    "            # ratio coexp = number of cell that express both divide by number of cell tha express at least one gene\n",
    "            nb_cell_exp.at[k[0],'ratio_coexp'] = nb_cell_exp.at[k[0],'exp_both'] / (nb_cell_exp.at[k[0],'exp_U']) \n",
    "        nb_cell_exp.at[k[0],'exp_host'] = g1.sum()\n",
    "        nb_cell_exp.at[k[0],'exp_nested'] = g2.sum()\n",
    "\n",
    "    return nb_cell_exp\n",
    "\n",
    "# nb_cell_expression(exp7)\n",
    "# nb_cell_expression(exp8)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26c56371-b2fe-4251-bbf3-7a8b68a71b13",
   "metadata": {},
   "source": [
    "#### iteration random and close pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "7f0faeb2-587c-4770-8b3b-2f2e95b33783",
   "metadata": {},
   "outputs": [],
   "source": [
    "### two way for caluculation of outlier that will be used later\n",
    "\n",
    "def outlier_up_zscore(ls):\n",
    "    z=stats.zscore(ls)\n",
    "    return sum(ls>3)\n",
    "\n",
    "def outlier_up_IQR(ls):\n",
    "    Q1=np.quantile(ls,.25)\n",
    "    Q3=np.quantile(ls,.75)\n",
    "    IQR=Q3-Q1\n",
    "    lim_up=Q3+(1.5*IQR)\n",
    "    return sum(ls>lim_up)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "769a15af-adc1-431d-b3c8-0c44d55a7750",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def list_exp_outlier_random(raw, exp, pairs='diff', it=1000, nb=567):\n",
    "    ls_mean_exp=[]\n",
    "    ls_exp_both=[]\n",
    "    ls_exp_U=[]\n",
    "    ls_out_up_z=[]\n",
    "    ls_out_up_iqr=[]\n",
    "    for i in range(it):\n",
    "        if pairs=='diff':\n",
    "            print('it',i,'diff')\n",
    "            pl = rand_pair_diff(raw, nb)\n",
    "        elif pairs=='close':\n",
    "            print('it',i,'close')\n",
    "            pl = rand_pair_close(raw, nb)\n",
    "        nb_cell_exp = nb_cell_expression(exp, pl)\n",
    "        coexp=nb_cell_exp['ratio_coexp'].astype('float64')\n",
    "        ls_mean_exp.append(np.mean(coexp))\n",
    "        ls_exp_both.append(sum(nb_cell_exp['exp_both']>0))\n",
    "        ls_exp_U.append(sum(nb_cell_exp['exp_U']>0))\n",
    "        ls_out_up_z.append(outlier_up_zscore(coexp))\n",
    "        ls_out_up_iqr.append(outlier_up_IQR(coexp))\n",
    "    \n",
    "    return pd.concat([pd.Series(ls_mean_exp,name='mean_exp'),\n",
    "                      pd.Series(ls_exp_both,name='exp_both'),\n",
    "                      pd.Series(ls_exp_U,name='exp_U'),\n",
    "                      pd.Series(ls_out_up_z,name='out_z'),\n",
    "                      pd.Series(ls_out_up_iqr,name='out_iqr')], axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7194eaf4-018d-4c62-a163-74f058fb0c81",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mean_exp</th>\n",
       "      <th>exp_both</th>\n",
       "      <th>exp_U</th>\n",
       "      <th>out_z</th>\n",
       "      <th>out_iqr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.081329</td>\n",
       "      <td>327</td>\n",
       "      <td>556</td>\n",
       "      <td>0</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.086069</td>\n",
       "      <td>329</td>\n",
       "      <td>558</td>\n",
       "      <td>0</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.095846</td>\n",
       "      <td>343</td>\n",
       "      <td>559</td>\n",
       "      <td>0</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.097414</td>\n",
       "      <td>348</td>\n",
       "      <td>552</td>\n",
       "      <td>0</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.091352</td>\n",
       "      <td>342</td>\n",
       "      <td>558</td>\n",
       "      <td>0</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>0.103700</td>\n",
       "      <td>335</td>\n",
       "      <td>556</td>\n",
       "      <td>0</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>0.084863</td>\n",
       "      <td>316</td>\n",
       "      <td>552</td>\n",
       "      <td>0</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>0.085159</td>\n",
       "      <td>346</td>\n",
       "      <td>559</td>\n",
       "      <td>0</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>0.104042</td>\n",
       "      <td>338</td>\n",
       "      <td>556</td>\n",
       "      <td>0</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>0.100789</td>\n",
       "      <td>352</td>\n",
       "      <td>559</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     mean_exp  exp_both  exp_U  out_z  out_iqr\n",
       "0    0.081329       327    556      0       77\n",
       "1    0.086069       329    558      0       51\n",
       "2    0.095846       343    559      0       67\n",
       "3    0.097414       348    552      0       62\n",
       "4    0.091352       342    558      0       63\n",
       "..        ...       ...    ...    ...      ...\n",
       "995  0.103700       335    556      0       67\n",
       "996  0.084863       316    552      0       58\n",
       "997  0.085159       346    559      0       67\n",
       "998  0.104042       338    556      0       67\n",
       "999  0.100789       352    559      0       61\n",
       "\n",
       "[1000 rows x 5 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "######## Generate data\n",
    "#df_list_diff_terminal=list_exp_outlier_random(raw,exp7,pairs='diff',it=1000)\n",
    "#df_list_close_terminal=list_exp_outlier_random(raw,exp7,pairs='close',it=1000)\n",
    "#df_list_diff_lineage=list_exp_outlier_random(raw,exp8,pairs='diff',it=1000)\n",
    "#df_list_close_lineage=list_exp_outlier_random(raw,exp8,pairs='close',it=1000)\n",
    "\n",
    "######## Save data\n",
    "#df_list_diff_terminal.to_csv('df_list_diff_exp7.csv')\n",
    "#df_list_close_terminal.to_csv('df_list_close_exp7.csv')\n",
    "#df_list_diff_lineage.to_csv('df_list_diff_exp8.csv')\n",
    "#df_list_close_lineage.to_csv('df_list_close_exp8.csv')\n",
    "\n",
    "######## Read data\n",
    "df_list_diff_terminal=pd.read_csv('df_list_diff_exp7.csv',index_col=0)\n",
    "df_list_close_terminal=pd.read_csv('df_list_close_exp7.csv',index_col=0)\n",
    "df_list_diff_lineage=pd.read_csv('df_list_diff_exp8.csv',index_col=0)\n",
    "df_list_close_lineage=pd.read_csv('df_list_close_exp8.csv',index_col=0)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0d33ee6-ac02-4ab9-a2d8-1d0b58c2e597",
   "metadata": {},
   "source": [
    "## 1/ la coexpression des paires emboitÃ©es nâ€™est pas rare."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "629ab7fe-ae01-4df8-8979-461272008729",
   "metadata": {},
   "source": [
    "## 1/ Coexpression of nested pairs is not rare"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddb50fd5-5f0b-4884-8e69-78a97cd4c7d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "print('######################','\\n', 'S8 Lineage')\n",
    "nes=pd.read_csv('/home/mateo/IGE/code_nested_gene_M/celltype_expression/cell_type_expression_nested_s8.csv',index_col=0)\n",
    "print(sum(nes['exp_both']>0),'nb de pairs ou il y a coexpression')\n",
    "print(sum(nes['exp_U']>0),'nb de pairs ou il y a expression')\n",
    "print(len(nes),'nb de pairs')\n",
    "print(sum(nes['exp_both']>0)/len(nes),'ratio nb de pairs cooexp/nb pairs total')\n",
    "print(sum(nes['exp_both']>0)/sum(nes['exp_U']>0),'ratio nb de pairs cooexp/nb de pairs exp')\n",
    "print('######################','\\n', 'S7 terminal cells')\n",
    "nes=pd.read_csv('/home/mateo/IGE/code_nested_gene_M/celltype_expression/cell_type_expression_nested_s7.csv',index_col=0)\n",
    "print(sum(nes['exp_both']>0),'nb de pairs ou il y a coexpression')\n",
    "print(sum(nes['exp_U']>0),'nb de pairs ou il y a expression')\n",
    "print(len(nes),'nb de pairs')\n",
    "print(sum(nes['exp_both']>0)/len(nes),'ratio nb de pairs cooexp/nb pairs total')\n",
    "print(sum(nes['exp_both']>0)/sum(nes['exp_U']>0),'ratio nb de pairs cooexp/nb de pairs exp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22039b59-3296-4086-aa01-291334b9ead1",
   "metadata": {},
   "outputs": [],
   "source": [
    "## faire la meme chose avec les donne random et les proches\n",
    "# iteration sur plusier pairs random proches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cecdc92-a4e8-4433-ad78-f714794a842f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_list\n",
    "n_iteration=10\n",
    "n_pairs=len(df_list)*566\n",
    "\n",
    "print(sum(df_list['both_diff']),' nb de pairs ou il y a coexpression')\n",
    "print(sum(df_list['U_diff']),' nb de pairs ou il y a expression' )\n",
    "print(n_pairs,'nb de pairs')\n",
    "print(sum(df_list['both_diff'])/n_pairs,'ratio nb de pairs cooexp/nb pairs total')\n",
    "print(sum(df_list['both_diff'])/sum(df_list['U_diff']),'ratio nb de pairs cooexp/nb de pairs exp')\n",
    "print('################')\n",
    "print(sum(df_list['both_close']),'nb de pairs ou il y a coexpression')\n",
    "print(sum(df_list['U_close']),'nb de pairs ou il y a expression')\n",
    "print(n_pairs,'nb de pairs')\n",
    "print(sum(df_list['both_close'])/n_pairs,'ratio nb de pairs cooexp/nb pairs total')\n",
    "print(sum(df_list['both_close'])/sum(df_list['U_close']),'ratio nb de pairs cooexp/nb de pairs exp')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fce5cf2c-87ec-4652-8b89-de7b710a4084",
   "metadata": {},
   "source": [
    "## 2/ les paires emboitÃ©es sont en moyenne moins coexprimÃ©es que des paires proches non-emboitÃ©es."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd871ab9-999d-4a97-84dc-ce3674d617c3",
   "metadata": {},
   "source": [
    "## 2/ Nested pairs are less coexpressed than non-nested random close pairs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82c200bd-e955-4e76-8b6d-1440e9e4735f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def graph_mean(selec):    \n",
    "    if selec=='S8':\n",
    "        ls=pd.read_csv(\"/home/mateo/IGE/code_nested_gene_M/list_mean_s8.csv\")\n",
    "        c=pd.read_csv('/home/mateo/IGE/code_nested_gene_M/celltype_expression/cell_type_expression_nested_s8.csv')['ratio_coexp']\n",
    "        n_mean=np.mean(c[c!=0])\n",
    "    if selec=='S7':\n",
    "        ls=pd.read_csv(\"/home/mateo/IGE/code_nested_gene_M/list_mean_s7.csv\")\n",
    "        c=pd.read_csv('/home/mateo/IGE/code_nested_gene_M/celltype_expression/cell_type_expression_nested_s7.csv')['ratio_coexp']\n",
    "        n_mean=np.mean(c[c!=0])\n",
    "        \n",
    "    a=ls['diff'] # ls['diff_mean'] soon to be updated\n",
    "    b=ls['close']\n",
    "    # print(a,b,c)\n",
    "    print(len(a))\n",
    "    print(stats.kstest(a,b))\n",
    "    \n",
    "    fig1 = px.histogram(a, range_x=[0, 1], color_discrete_sequence=['#FECB52'], opacity=1)\n",
    "    fig2 = px.histogram(b, range_x=[0, 1], color_discrete_sequence=['#EF553B'], opacity=1)\n",
    "\n",
    "    fig=go.Figure()\n",
    "\n",
    "    fig.add_trace(fig1.data[0])\n",
    "    fig.add_trace(fig2.data[0])\n",
    "    \n",
    "    fig.update_traces(xbins=dict(start=0, end=1, size=.001))\n",
    "    \n",
    "    fig.add_vline(x=n_mean, line_width=2, line_color='#636EFA', annotation_text='nested mean', annotation_position='top left')\n",
    "\n",
    "    fig.update_xaxes(title_text=\"mean frequency of coexpression\", range=[0, .5])\n",
    "    fig.update_yaxes(title_text=\"number of dataset\")\n",
    "\n",
    "    return fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9272b800-9d58-46a4-b1f7-8a2a65f63349",
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_mean('S7')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba8c2001-1c09-441d-836f-0ea6b9c0c0ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ls=pd.read_csv(\"/home/mateo/IGE/code_nested_gene_M/list_mean_s7.csv\")\n",
    "# a=ls['diff'] # ls['diff_mean'] soon to be updated\n",
    "c=pd.read_csv('/home/mateo/IGE/code_nested_gene_M/celltype_expression/cell_type_expression_nested_s7.csv')['ratio_coexp']\n",
    "n_mean=np.mean(c[c!=0])\n",
    "\n",
    "a=df_list['mean_diff']\n",
    "a_mean=np.mean(a)\n",
    "\n",
    "def test_stat_1(distrib,value):\n",
    "    # One-sample t-test\n",
    "    t_statistic, p_value_t = stats.ttest_1samp(distrib, value)\n",
    "    print(\"One-sample t-test:\")\n",
    "    print(\"t-statistic:\", t_statistic)\n",
    "    print(\"p-value:\", p_value_t)\n",
    "    # Wilcoxon signed-rank test\n",
    "    w_statistic, p_value_w = stats.wilcoxon(distrib - value)\n",
    "    print(\"\\nWilcoxon signed-rank test:\")\n",
    "    print(\"W-statistic:\", w_statistic)\n",
    "    print(\"p-value:\", p_value_w)\n",
    "    return True\n",
    "\n",
    "test_stat_1(a,n_mean)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24f63e40-9a11-473f-b47f-12a52c15610d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## test statistique\n",
    "# print(\"p-value test Kolmogorov-Smirnov\" ,stats.kstest(df['ratio_coexp'], nes['ratio_coexp'])[1])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "423e3aab-ed48-4556-b3fa-9d49c3245d3a",
   "metadata": {},
   "source": [
    "## 3/ il y a chez les paires emboitÃ©es des outliers trÃ¨s fortement coexprimÃ©s. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b0b3d20-e967-4ddc-bf2e-7c37791a3360",
   "metadata": {},
   "source": [
    "## 3/ Nested pairs contain outliers with high coexpression"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e73c633-1cb9-4e04-9458-69263644d6c0",
   "metadata": {},
   "source": [
    "### probleme : precedent calcule fait su les moyenne pas les donnee de coexp donc refaire outlier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e34b2de0-374c-488a-ab49-5fb3447762b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#diff_S7=pd.read_csv('/home/mateo/IGE/code_nested_gene_M/coexp_mean/list_mean_s7_diff.csv',index_col=0)\n",
    "#diff_S8=pd.read_csv('/home/mateo/IGE/code_nested_gene_M/coexp_mean/list_mean_s8_diff.csv',index_col=0)\n",
    "#close_S7=pd.read_csv('/home/mateo/IGE/code_nested_gene_M/coexp_mean/list_mean_s7_close.csv',index_col=0)\n",
    "#close_S8=pd.read_csv('/home/mateo/IGE/code_nested_gene_M/coexp_mean/list_mean_s8_close.csv',index_col=0)\n",
    "\n",
    "#close_S7_conv=pd.read_csv('/home/mateo/IGE/code_nested_gene_M/coexp_mean/list_mean_s7_conv.csv',index_col=0)\n",
    "#close_S7_div=pd.read_csv('/home/mateo/IGE/code_nested_gene_M/coexp_mean/list_mean_s7_div.csv',index_col=0)\n",
    "#close_S8_conv=pd.read_csv('/home/mateo/IGE/code_nested_gene_M/coexp_mean/list_mean_s8_conv.csv',index_col=0)\n",
    "#close_S8_div=pd.read_csv('/home/mateo/IGE/code_nested_gene_M/coexp_mean/list_mean_s8_div.csv',index_col=0)\n",
    "\n",
    "#exp_nested_S7=pd.read_csv('/home/mateo/IGE/code_nested_gene_M/celltype_expression/cell_type_expression_nested_s7.csv')['ratio_coexp']\n",
    "#exp_nested_S8=pd.read_csv('/home/mateo/IGE/code_nested_gene_M/celltype_expression/cell_type_expression_nested_s8.csv')['ratio_coexp']\n",
    "\n",
    "#print(diff_S7)\n",
    "#px.box(diff_S7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ed8e1fe-c98e-4ea8-b246-e7a4f48c084f",
   "metadata": {},
   "outputs": [],
   "source": [
    "### function to get the value of outlier z1 and z3\n",
    "\n",
    "def get_z_val(values):\n",
    "    m = np.mean(values)\n",
    "    sd = np.std(values)\n",
    "    zd3 = m-(3*sd)\n",
    "    zd1 = m-(1*sd)\n",
    "    zu1 = m+(1*sd)\n",
    "    zu3 = m+(3*sd)\n",
    "    return zd3,zd1,zu1,zu3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "095ccf77-b2d0-49b9-a6e1-404b20e80c9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# fig3 = px.histogram(c, range_x=[0, 1], color_discrete_sequence=['#636EFA'], barmode='overlay', opacity=0.7, marginal=\"box\")\n",
    "c=pd.read_csv('/home/mateo/IGE/code_nested_gene_M/celltype_expression/cell_type_expression_nested_s7.csv')['ratio_coexp']\n",
    "diff=rand_pair_diff(raw)\n",
    "close=rand_pair_close(raw)\n",
    "a=nb_cell_expression(exp7,diff)['ratio_coexp'].astype('float64')\n",
    "b=nb_cell_expression(exp7,close)['ratio_coexp'].astype('float64')\n",
    "\n",
    "df=pd.DataFrame({'diff':a,'close':b,'nested':c})\n",
    "fig=px.box(a,color_discrete_sequence=[\"#FECB52\"])\n",
    "fig.add_trace(px.box(b,color_discrete_sequence=[\"#EF553B\"]).data[0])\n",
    "fig.add_trace(px.box(c,color_discrete_sequence=[\"#636EFA\"]).data[0])\n",
    "fig.show()\n",
    "\n",
    "df =pd.DataFrame(dict(\n",
    "    series=np.concatenate(([\"nested pair\"]*len(a), [\"random pair on different chromosome\"]*len(b), [\"random pair at 1kb\"]*len(c))), \n",
    "    data  =np.concatenate((c,a,b))\n",
    "))\n",
    "\n",
    "fig= px.box(df, x=\"data\", color=\"series\", color_discrete_sequence=[\"#636EFA\",\"#FECB52\",\"#EF553B\"],\n",
    "                  labels={\"data\" : \"value of coexpresion\",\n",
    "                          \"series\" : \"type of pair\"\n",
    "                      }).update_layout(yaxis_title=\"number of pair\")\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "306fd8a6-c045-4139-b265-c2b24a670883",
   "metadata": {},
   "outputs": [],
   "source": [
    "# fig3 = px.histogram(c, range_x=[0, 1], color_discrete_sequence=['#636EFA'], barmode='overlay', opacity=0.7, marginal=\"box\")\n",
    "c=pd.read_csv('/home/mateo/IGE/code_nested_gene_M/celltype_expression/cell_type_expression_nested_s7.csv')['ratio_coexp']\n",
    "diff=rand_pair_diff(raw)\n",
    "close=rand_pair_close(raw)\n",
    "a=nb_cell_expression(exp7,diff)['ratio_coexp'].astype('float64')\n",
    "b=nb_cell_expression(exp7,close)['ratio_coexp'].astype('float64')\n",
    "\n",
    "#a=a[a!=0]\n",
    "#b=b[b!=0]\n",
    "#c=c[c!=0]\n",
    "\n",
    "up_a=get_z_val(a)[3]\n",
    "up_b=get_z_val(b)[3]\n",
    "up_c=get_z_val(c)[3]\n",
    "\n",
    "fig=px.box(x=a,color_discrete_sequence=[\"#FECB52\"])\n",
    "fig.add_vline(x=up_a,line_color=\"#FECB52\")\n",
    "fig.add_vline(x=up_b,line_color=\"#EF553B\")\n",
    "fig.add_vline(x=up_c,line_color=\"#636EFA\")\n",
    "fig.add_trace(px.box(x=b,color_discrete_sequence=[\"#EF553B\"]).data[0])\n",
    "fig.add_trace(px.box(x=c,color_discrete_sequence=[\"#636EFA\"]).data[0])\n",
    "fig.show()\n",
    "\n",
    "df =pd.DataFrame(dict(\n",
    "    series=np.concatenate(([\"nested pair\"]*len(a), [\"random pair on different chromosome\"]*len(b), [\"random pair at 1kb\"]*len(c))), \n",
    "    data  =np.concatenate((c,a,b))\n",
    "))\n",
    "\n",
    "fig= px.box(df, x=\"data\", color=\"series\", color_discrete_sequence=[\"#636EFA\",\"#FECB52\",\"#EF553B\"],\n",
    "                  labels={\"data\" : \"value of coexpresion\",\n",
    "                          \"series\" : \"type of pair\"\n",
    "                      }).update_layout(yaxis_title=\"number of pair\")\n",
    "fig.add_vline(x=up_a,line_color=\"#FECB52\")\n",
    "fig.add_vline(x=up_b,line_color=\"#EF553B\")\n",
    "fig.add_vline(x=up_c,line_color=\"#636EFA\")\n",
    "fig.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25a4a318-b4bb-4c85-9d23-92b6725f2757",
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(a,b,c)\n",
    "pour_vincent=pd.DataFrame({'random_diff' : a, 'random_close' : b, 'nested' : c})\n",
    "pour_vincent.to_csv('pour_vincent.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47a58ab2-08ca-45e9-9e0e-969ff24e50d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"p-value test Kolmogorov-Smirnov\" ,stats.kstest(a,c)[1])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f42ddae-eccf-473d-b059-f80fda189ce2",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# a=nb_cell_expression(exp7,diff)['ratio_coexp'].astype('float64')\n",
    "# b=nb_cell_expression(exp7,close)['ratio_coexp'].astype('float64')\n",
    "za=stats.zscore(a)\n",
    "zb=stats.zscore(b)\n",
    "zc=stats.zscore(c)\n",
    "\n",
    "up_a=a[za[za==min(za[za>3])].index[0]]\n",
    "if len(zb[zb>3]) > 0: \n",
    "    up_b=b[zb[zb==min(zb[zb>3])].index[0]]\n",
    "else :\n",
    "    up_b=1\n",
    "up_c=c[zc[zc==min(zc[zc>3])].index[0]]\n",
    "print(up_a,up_b,up_c)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "277bd0c7-c003-43ab-a720-bb2e6aaf82dc",
   "metadata": {},
   "source": [
    "### doit tester avec l'go Q3+1.5 IQR pour compter les outlier "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68d6c7c2-415f-42c5-bdc0-627728740169",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig=px.box(x=a,color_discrete_sequence=[\"#FECB52\"])\n",
    "fig.add_vline(x=up_a,line_color=\"#FECB52\")\n",
    "fig.show()\n",
    "fig=px.box(x=b,color_discrete_sequence=[\"#EF553B\"])\n",
    "fig.add_vline(x=up_b,line_color=\"#EF553B\")\n",
    "fig.show()\n",
    "fig=px.box(x=c,color_discrete_sequence=[\"#636EFA\"])\n",
    "fig.add_vline(x=up_c,line_color=\"#636EFA\")\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d87d724e-3312-40a4-b183-1908a99a5e7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "azd3,azd1,azu1,azu3=get_z_val(a)\n",
    "bzd3,bzd1,bzu1,bzu3=get_z_val(b)\n",
    "czd3,czd1,czu1,czu3=get_z_val(c)\n",
    "\n",
    "fig = make_subplots(rows=3, cols=1,\n",
    "                    shared_xaxes=True,\n",
    "                    vertical_spacing=0.02)\n",
    "fig1=go.Figure()\n",
    "fig1.add_trace(go.Box(x=[a], name='diff', marker_color=\"#FECB52\", boxpoints='outliers',))\n",
    "fig1.update_traces(q1=[azd1], median=[np.median(a)],\n",
    "                  q3=[azu1], lowerfence=[azd3],\n",
    "                  upperfence=[azu3], mean=[np.mean(a)])\n",
    "fig2 = go.Figure()\n",
    "fig2.add_trace(go.Box(x=[b], name='close', marker_color=\"#EF553B\", boxpoints='outliers'))\n",
    "fig2.update_traces(q1=[bzd1], median=[np.median(b)],\n",
    "                  q3=[bzu1], lowerfence=[bzd3],\n",
    "                  upperfence=[bzu3], mean=[np.mean(b)])\n",
    "fig3 = go.Figure()\n",
    "fig3.add_trace(go.Box(x=[c], name='nested', marker_color=\"#636EFA\", boxpoints='outliers'))\n",
    "fig3.update_traces(q1=[czd1], median=[np.median(c)],\n",
    "                  q3=[czu1], lowerfence=[czd3],\n",
    "                  upperfence=[czu3], mean=[np.mean(c)])\n",
    "\n",
    "fig.add_trace(fig1.data[0],row=1, col=1)\n",
    "fig.add_trace(fig2.data[0],row=2, col=1)\n",
    "fig.add_trace(fig3.data[0],row=3, col=1)\n",
    "fig.show()\n",
    "fig.update(layout_xaxis_range = [0,1.1])\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55457457-44a8-4b57-9357-946d0b9b4a98",
   "metadata": {},
   "outputs": [],
   "source": [
    "#fig.add_trace(go.Box(x=b,name='close'))\n",
    "#fig.add_trace(go.Box(x=c,name='nested'))\n",
    "#fig.update_traces(q1=[azd1,bzd1,czd1], median=[np.median(a),np.median(b),np.median(c)],\n",
    "#                  q3=[azu1,bzu1,czu1], lowerfence=[azd3,bzd3,czd3],\n",
    "#                  upperfence=[azu3,bzu3,czu3], mean=[np.mean(a),np.mean(b),np.mean(c)],\n",
    "#                  sd=[np.std(a),np.std(b),np.std(c)])\n",
    "#\n",
    "#fig.update_traces(q1=[azd1], median=[np.median(a)],\n",
    "#                  q3=[azu1], lowerfence=[azd3],\n",
    "#                  upperfence=[azu3], mean=[np.mean(a)],\n",
    "#                  sd=[np.std(a)])\n",
    "#fig.update_traces(q1=[bzd1], median=[np.median(b)],\n",
    "#                  q3=[bzu1], lowerfence=[bzd3],\n",
    "#                  upperfence=[bzu3], mean=[np.mean(b)],\n",
    "#                  sd=[np.std(b)])\n",
    "#fig.update_traces(q1=[czd1], median=[np.median(c)],\n",
    "#                  q3=[czu1], lowerfence=[czd3],\n",
    "#                  upperfence=[czu3], mean=[np.mean(c)],\n",
    "#                  sd=[np.std(c)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31fd0723-9d27-4218-9db8-ccdb98e0503d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def outlier_up_IQR(ls):\n",
    "    Q1=np.quantile(ls,.25)\n",
    "    Q3=np.quantile(ls,.75)\n",
    "    IQR=Q3-Q1\n",
    "    out_up=Q3+(1.5*IQR)\n",
    "    return out_up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "024430fe-870f-4b26-a286-c1fe17cd22fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_up=outlier_up_IQR(a)\n",
    "print(sum(a>out_up))\n",
    "out_up=outlier_up_IQR(b)\n",
    "print(sum(b>out_up))\n",
    "out_up=outlier_up_IQR(c)\n",
    "print(sum(c>out_up))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c711ef2-11f8-44e1-bd76-388dd1695b53",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "it=100\n",
    "nb=566\n",
    "alpha=0\n",
    "out_up=[]\n",
    "for i in range(it):\n",
    "    print('it',i,'diff')\n",
    "    pl = rand_pair_diff(raw, nb)\n",
    "    #pl = rand_pair_close(raw, 'both', nb)\n",
    "    a = nb_cell_expression(exp7, pl, alpha)['ratio_coexp'].astype('float64')\n",
    "    lim=outlier_up_IQR(a)\n",
    "    out_up.append(sum(a>lim))\n",
    "\n",
    "print(np.mean(out_up))\n",
    "fig=px.histogram(out_up)\n",
    "fig.add_vline(x=46)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "448516b8-4d1e-45af-b5b6-9ac44cd14533",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compter avec les IQR \n",
    "# refaire le plot avec les Z3 a la palce des IQR"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5261e3fd-cbd0-4be0-8221-27de8930a2fc",
   "metadata": {},
   "source": [
    "### outlier with zscore refaire mieux"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5dfa16ff-f2d4-4baf-b794-7631d4e1fd6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def out_z(z,tresh=3):\n",
    "    print(sum(z>tresh),'outlier up')\n",
    "    print(sum(z<-tresh),\"outlier down\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13686b34-d39e-4200-b53a-e93caec5f603",
   "metadata": {},
   "outputs": [],
   "source": [
    "za=stats.zscore(a)\n",
    "zb=stats.zscore(b)\n",
    "zc=stats.zscore(c)\n",
    "for i in [za,zb,zc]:\n",
    "    print('###')\n",
    "    out_z(i)\n",
    "print('############')\n",
    "for i in [za,zb,zc]:\n",
    "    print('###')\n",
    "    out_z(i,1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d82ae6ab-dcb0-4fa8-b26e-984976f0b149",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "it=100\n",
    "nb=566\n",
    "alpha=0\n",
    "out_up=[]\n",
    "for i in range(it):\n",
    "    print('it',i,'diff')\n",
    "    pl = rand_pair_diff(raw, nb)\n",
    "    #pl = rand_pair_close(raw, 'both', nb)\n",
    "    a = nb_cell_expression(exp7, pl, alpha)['ratio_coexp'].astype('float64')\n",
    "    a=a[a!=0]\n",
    "    za=stats.zscore(a)\n",
    "    out_up.append(sum(za>3))\n",
    "\n",
    "print(np.mean(out_up))\n",
    "fig=px.histogram(out_up)\n",
    "fig.add_vline(x=12)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32b0221d-d93c-4086-9c70-2e054144b84b",
   "metadata": {},
   "outputs": [],
   "source": [
    "c7, c8 = exp_nested_S7, exp_nested_S8\n",
    "zc7=stats.zscore(c7)\n",
    "zc8=stats.zscore(c8)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8072d8ec-ba30-48e4-9f39-dec034f1ea25",
   "metadata": {},
   "source": [
    "In this article, we have shown four statistical procedures of detecting outliers in the dataset. For normal dataset, IQR multiplier of 1.7 is similar to stdev multiplier of 3 which is also similar to setting cut-off Z-score at 3 as detection limit. Statisticians use modified Z-score to minimize the influence of outliers on Z-score. This modified Z-score indicates the relative strength of the outlier and how much it deviated from the Z-score it was supposed to have. All these procedures are standard procedures to determine outliers statistically.\n",
    "\n",
    "https://towardsdatascience.com/outlier-detection-part1-821d714524c\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ebf4b21-cded-4aca-8b06-142925fe7dff",
   "metadata": {},
   "source": [
    "### extract outlier as list "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b11c7b2-4262-4bc7-b90e-7518b1dbbc30",
   "metadata": {},
   "outputs": [],
   "source": [
    "nes_7=pd.read_csv('/home/mateo/IGE/code_nested_gene_M/celltype_expression/cell_type_expression_nested_s7.csv',index_col=0)\n",
    "nes_8=pd.read_csv('/home/mateo/IGE/code_nested_gene_M/celltype_expression/cell_type_expression_nested_s8.csv',index_col=0)\n",
    "ls_name=pd.read_csv('/home/mateo/IGE/code_nested_gene_M/pair_list/nested_pair_name.csv',index_col=0)\n",
    "def extract_outlier(z,nes,tresh=3,ls_name=ls_name):\n",
    "    #print(sum(z>tresh))\n",
    "    id_out = nes.iloc[z[z>tresh].index,0:2].index\n",
    "    outliers = ls_name.iloc[id_out,:]\n",
    "    return outliers\n",
    "\n",
    "extract_outlier(zc7,nes_7)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ca337e4-15fe-4c7c-b2c6-a395fbb11b43",
   "metadata": {},
   "source": [
    "## 4/ la probabilitÃ© dâ€™avoir une expression de lâ€™hote quand lâ€™interne sâ€™exprime est plus Ã©levÃ©e que lâ€™inverse"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3e1a27b-3288-4658-a56f-7c62fdd6c8c3",
   "metadata": {},
   "source": [
    "## 4/ Probability of expressing the host gene, when the nested is expressed, is higher than the opposite"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bc1ceba-046a-417e-91f9-8604612572ef",
   "metadata": {},
   "source": [
    "### nes/profil coexp different Ã  eclaircir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9370ece0-e159-400e-8a14-49877c8576a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "profil_s8=pd.read_pickle('/home/mateo/IGE/code_nested_gene_M/celltype_profil_s8.pkl')\n",
    "profil_s7=pd.read_pickle('/home/mateo/IGE/code_nested_gene_M/celltype_profil_s7.pkl')\n",
    "\n",
    "def prob_of_expression(profil):\n",
    "    profil=profil.drop(len(profil)-1)\n",
    "    B=profil['propB']\n",
    "    H=profil['propH']\n",
    "    N=profil['propN']\n",
    "    l=[]\n",
    "    l2=[]\n",
    "    for i in range(len(profil)):\n",
    "        if B[i]+N[i] > 0:\n",
    "            l.append(B[i]/(B[i]+N[i]))\n",
    "        if B[i]+H[i] > 0:\n",
    "            l2.append(B[i]/(B[i]+H[i]))\n",
    "\n",
    "    fig=px.histogram(l, histnorm='percent', range_x=[0,1]).update_layout(yaxis_title=\"percentage of pair\",xaxis_title=\"probability of expressing host if nested is expressed\")\n",
    "    fig.add_vline(x=sum(l)/len(l), line_width=2, line_color='red', annotation_text='means sum(prob) / nb prob', annotation_position='top right')\n",
    "    fig.show()\n",
    "    print(sum(l)/len(l), 'mean of probability host if nested expressed')\n",
    "\n",
    "    fig2=px.histogram(l2, histnorm='percent', range_x=[0,1]).update_layout(yaxis_title=\"percentage of pair\",xaxis_title=\"probability of expressing nested if host is expressed\")\n",
    "    fig2.add_vline(x=sum(l2)/len(l2), line_width=2, line_color='red', annotation_text='means sum(prob) / nb prob', annotation_position='top right')\n",
    "    fig2.show()\n",
    "    print(sum(l2)/len(l2), 'mean of probability nested if host expressed')\n",
    "\n",
    "prob_of_expression(profil_s7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8aaa757-8b5d-436b-a513-bb619efb85dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def prob_of_expression2(nes):\n",
    "    B=nes['exp_both']\n",
    "    H=nes['exp_host']\n",
    "    N=nes['exp_nested']\n",
    "    l=[]\n",
    "    l2=[]\n",
    "    for i in range(len(nes)):\n",
    "        if B[i]+N[i] > 0:\n",
    "            l.append(B[i]/(B[i]+N[i]))\n",
    "        if B[i]+H[i] > 0:\n",
    "            l2.append(B[i]/(B[i]+H[i]))\n",
    "\n",
    "    fig=px.histogram(l,range_x=[0,1]).update_layout(yaxis_title=\"number of pair\",xaxis_title=\"probability of expressing host if nested is expressed\")\n",
    "    fig.add_vline(x=sum(l)/len(l), line_width=2, line_color='red', annotation_text='means sum(prob) / nb prob', annotation_position='top left')\n",
    "    fig.update_layout()\n",
    "    fig.show()\n",
    "    print(sum(l)/len(l), 'mean of probability host if nested expressed')\n",
    "\n",
    "    fig2=px.histogram(l2,range_x=[0,1]).update_layout(yaxis_title=\"number of pair\",xaxis_title=\"probability of expressing nested if host is expressed\")\n",
    "    fig2.add_vline(x=sum(l2)/len(l2), line_width=2, line_color='red', annotation_text='means sum(prob) / nb prob', annotation_position='top left')\n",
    "    fig2.show()\n",
    "    print(sum(l2)/len(l2), 'mean of probability nested if host expressed')\n",
    "    \n",
    "prob_of_expression2(nes_7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bd7b73e-4f6b-4ba5-b613-b5a1e0545c8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#WS285=pd.read_csv('/home/mateo/Downloads/mart_exportWS285.txt')\n",
    "#raw_ex=pd.read_csv('/home/mateo/IGE/code_nested_gene_M/pair_list/mart_export_exon.txt')\n",
    "#raw=pd.read_csv(file_mart , delimiter=(\",\"))\n",
    "#print(len(np.unique(raw['Gene start (bp)'])))\n",
    "#print(len(np.unique(WS285['Gene start (bp)'])))\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
